src/v5/client_commands.rs

```diff
@@ -0,0 +1,250 @@
+use byteorder::{ByteOrder, NetworkEndian};
+#[cfg(not(feature = "tokio"))]
+use futures::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use thiserror::Error;
+#[cfg(feature = "tokio")]
+use tokio_compat::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+
+use crate::v5::{
+    SocksV5AddressType, SocksV5Command, SocksV5Host, SocksV5RequestError, SocksV5RequestStatus,
+};
+use crate::SocksVersion;
+
+/// Writes a SOCKSv5 request with the specified command, host and port.
+///
+/// # Errors
+///
+/// If writing to `writer` fails, this function will return the I/O error.
+///
+/// # Panics
+///
+/// If `host` is a domain name, the length of which is greater than 255 bytes.
+/// The SOCKSv5 specification leaves only a single octet for encoding the domain name length,
+/// so a target longer than 255 bytes cannot be properly encoded.
+pub async fn write_request<Writer>(
+    mut writer: Writer,
+    command: SocksV5Command,
+    host: SocksV5Host,
+    port: u16,
+) -> std::io::Result<()>
+where
+    Writer: AsyncWrite + Unpin,
+{
+    let mut data = Vec::<u8>::with_capacity(
+        6 + match &host {
+            SocksV5Host::Domain(domain) => {
+                assert!(
+                    domain.len() <= 256,
+                    "domain name must be shorter than 256 bytes"
+                );
+                1 + domain.len()
+            }
+            SocksV5Host::Ipv4(_) => 4,
+            SocksV5Host::Ipv6(_) => 16,
```

Comment:
This should be a method: `SocksV5Host::repr_len()` or something like this.

Reply:
Yes, that's a good idea. I'm focused more on features than this kind of small refactorings right now, but at some point it would be nice to come back to this.

Do you think such methods should be `pub` or private?

Reply:
I'd say private for now,.